glean/src/core/metrics/types/labeled.ts

```diff
@@ -0,0 +1,174 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { CommonMetricData, MetricType } from "..";
+import Glean from "../../glean";
+import CounterMetricType from "./counter";
+import BooleanMetricType from "./boolean";
+import StringMetricType from "./string";
+
+const MAX_LABELS = 16;
+const OTHER_LABEL = "__other__";
+const MAX_LABEL_LENGTH = 61;
+
+// ** IMPORTANT **
+// When changing this documentation or the regex, be sure to change the same code
+// in the Glean SDK repository as well.
+//
+// This regex is used for matching against labels and should allow for dots,
+// underscores, and/or hyphens. Labels are also limited to starting with either
+// a letter or an underscore character.
+//
+// Some examples of good and bad labels:
+//
+// Good:
+// * `this.is.fine`
+// * `this_is_fine_too`
+// * `this.is_still_fine`
+// * `thisisfine`
+// * `_.is_fine`
+// * `this.is-fine`
+// * `this-is-fine`
+// Bad:
+// * `this.is.not_fine_due_tu_the_length_being_too_long_i_thing.i.guess`
+// * `1.not_fine`
+// * `this.$isnotfine`
+// * `-.not_fine`
+const LABEL_REGEX = /^[a-z_][a-z0-9_-]{0,29}(\.[a-z_][a-z0-9_-]{0,29})*$/;
+
+type SupportedLabeledTypes = CounterMetricType | BooleanMetricType | StringMetricType;
+
+class LabeledMetricType<T extends SupportedLabeledTypes> {
+  // Define an index signature to make the Proxy aware of the expected return type.
+  // Note that this is required because TypeScript does not allow different input and
+  // output types in Proxy (https://github.com/microsoft/TypeScript/issues/20846). 
+  [label: string]: T;
+
+  constructor(
+    meta: CommonMetricData,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    submetric: new (...args: any) => T,
+    labels?: string[],
+  ) {
+    return new Proxy(this, {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      get: (_target: LabeledMetricType<T>, label: string): any => {
```

Comment:
```suggestion
      get: (_target: LabeledMetricType<T>, label: string): T => {
```

We need to be mindful of returning `any` here specially because it is the user facing API.